lua/roslyn/sln/utils.lua

```diff
@@ -82,61 +113,65 @@ function M.root(buffer)
         }
     end
 
+
     if broad_search then
         local git_root = vim.fs.root(buffer, ".git")
         local search_root = git_root and sln:match(git_root) and git_root or sln
 
-        local solutions = find_solutions(search_root)
+        local solutions, solution_filters = find_solutions(search_root)
 
         return {
             solutions = solutions,
+            solution_filters = solution_filters,
             projects = projects,
         }
     else
+        local slnf = targets.slnf_dir
+
         return {
             solutions = find_files_with_extension(sln, ".sln"),
+            solution_filters = slnf and find_files_with_extension(slnf, ".slnf"),
             projects = projects,
         }
     end
 end
 
----Tries to predict which solutions to use if we found some
----returning the potentially predicted solution
----Notifies the user if we still have multiple to choose from
+---Tries to predict which target to use if we found some
+---returning the potentially predicted target
 ---@param root RoslynNvimRootDir
----@return string?
-function M.predict_sln_file(root)
+---@return boolean multiple, string? predicted_target
+function M.predict_target(root)
     if not root.solutions then
-        return nil
+        return false, nil
     end
 
     local config = require("roslyn.config").get()
-    local solutions = vim.iter(root.solutions)
-        :filter(function(solution)
-            if config.ignore_sln and config.ignore_sln(solution) then
+    local sln_api = require("roslyn.sln.api")
+
+    local filtered_targets = vim.iter({ root.solutions, root.solution_filters })
+        :flatten()
+        :filter(function(target)
+            if config.ignore_target and config.ignore_target(target) then
                 return false
             end
+
             return not root.projects
                 or vim.iter(root.projects.files):any(function(csproj_file)
-                    return require("roslyn.sln.api").exists_in_solution(solution, csproj_file)
+                    return sln_api.exists_in_target(target, csproj_file)
                 end)
         end)
         :totable()
 
-    if #solutions > 1 then
-        local chosen = config.choose_sln and config.choose_sln(solutions) or nil
+    if #filtered_targets > 1 then
+        local chosen = config.choose_target and config.choose_target(filtered_targets)
+
         if chosen then
-            return chosen
+            return false, chosen
         end
 
-        vim.notify(
```

Comment:
I think I maybe prefer having this inside here and having this method just return a single value

Reply:
The idea was the early return - see my comments [here](https://github.com/seblj/roslyn.nvim/pull/128#issuecomment-2560311938) and [here](https://github.com/seblj/roslyn.nvim/pull/128#discussion_r1896153094). The motivation for https://github.com/seblj/roslyn.nvim/commit/bc291551a637204088dc3a8d125eb424fbf0b41a was

>  Its extremely rare that I want to fall back to a project instead of choosing between potential targets myself, so I thought I'd suggest it.

But if `lock_target = false` this would ask the user on every file, so it probably only makes sense when `lock_target = true`. Therefore, I'd change it to

```lua
if multiple and roslyn_config.lock target then
    vim.notify(...)
    return -- Don't start the LSP from the nearest project,  make the user pick a targer
end
```

That said, I'm happy to revert bc291551a637204088dc3a8d125eb424fbf0b41a and discuss separate to this PR.

Reply:
Ahaa yeah okay that makes sense! Then we can just leave it like this😄